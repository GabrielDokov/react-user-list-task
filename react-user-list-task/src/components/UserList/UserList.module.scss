.title {
  margin: 0;
}

.buttonsContainer {
  margin: 24px 0 0;
  gap: 12px;
}
