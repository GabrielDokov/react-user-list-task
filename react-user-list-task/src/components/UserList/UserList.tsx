import { useAppSelector } from "../../store";
import { Collapse, Typography, Button, Spin, Flex } from "antd";
import EditUserForm from "../EditUserForm/EditUserForm";
import { useState } from "react";
import { useNavigate } from "react-router";
import { UserData } from "../../types/UserData";
import classes from "./UserList.module.scss";
import UserCard from "../UserCard/UserCard";
import { useTranslation } from "react-i18next";

const UserList = () => {
  const [editingUser, setEditingUser] = useState<UserData | null>(null);
  const { users, isLoading } = useAppSelector((state) => state.userInfo);
  const navigate = useNavigate();
  const { t } = useTranslation();

  return (
    <>
      {editingUser && (
        <EditUserForm user={editingUser} isFormOpen={true} onClose={() => setEditingUser(null)} />
      )}
      {isLoading ? (
        <Spin fullscreen spinning={isLoading} />
      ) : (
        <>
          <Typography.Title className={classes.title}>{t("titles.users")}</Typography.Title>
          <Collapse
            style={{ width: 800 }}
            accordion
            items={users.map((user) => ({
              key: user.id,
              label: user.name,
              children: (
                <UserCard user={user}>
                  <Flex className={classes.buttonsContainer}>
                    <Button type="primary" onClick={() => setEditingUser(user)}>
                      {t("buttons.edit")}
                    </Button>
                    <Button type="primary" onClick={() => navigate(`/posts/${user.id}`)}>
                      {t("buttons.seePosts")}
                    </Button>
                  </Flex>
                </UserCard>
              ),
            }))}
          />
        </>
      )}
    </>
  );
};

export default UserList;
