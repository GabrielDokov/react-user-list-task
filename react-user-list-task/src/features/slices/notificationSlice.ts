import { createSlice, isAnyOf } from "@reduxjs/toolkit";
import { fetchTasksThunk } from "../thunks/fetchTasksThunk";
import { fetchPostsByUserIdThunk } from "../thunks/fetchPostsByUserIdThunk";
import { fetchUsersThunk } from "../thunks/fetchUsersThunk";
import { updatePostThunk } from "../thunks/updatePostThunk";
import { deletePostByIdThunk } from "../thunks/deletePostByIdThunk";

export type NotificationState = {
  hasError: boolean;
  message: string;
};

const initialState: NotificationState = {
  hasError: false,
  message: "",
};

const notificationSlice = createSlice({
  name: "notification",
  initialState,
  reducers: {
    resetError: (state) => {
      state.hasError = false;
      state.message = "";
    },
  },
  extraReducers: ({ addMatcher }) => {
    addMatcher(
      isAnyOf(
        fetchUsersThunk.rejected,
        fetchTasksThunk.rejected,
        fetchPostsByUserIdThunk.rejected,
        updatePostThunk.rejected,
        deletePostByIdThunk.rejected,
      ),
      (state, action) => {
        state.hasError = true;
        if (action.payload) {
          state.message = action.payload;
        }
      },
    );
  },
});

export const { resetError } = notificationSlice.actions;

export default notificationSlice.reducer;
