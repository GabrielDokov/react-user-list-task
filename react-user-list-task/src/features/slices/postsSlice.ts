import { createSlice, isAnyOf, PayloadAction } from "@reduxjs/toolkit";
import { PostData } from "../../types/PostsData";
import { fetchPostsByUserIdThunk } from "../thunks/fetchPostsByUserIdThunk";
import { deletePostByIdThunk } from "../thunks/deletePostByIdThunk";
import { updatePostThunk } from "../thunks/updatePostThunk";

export type PostsState = {
  posts: PostData[];
  isLoading: boolean;
};

const initialState: PostsState = {
  posts: [],
  isLoading: false,
};

const postsSlice = createSlice({
  name: "posts",
  initialState,
  reducers: {},
  extraReducers: ({ addCase, addMatcher }) => {
    addCase(fetchPostsByUserIdThunk.fulfilled, (state, action: PayloadAction<PostData[]>) => {
      state.posts = action.payload;
      state.isLoading = false;
    });
    addCase(deletePostByIdThunk.fulfilled, (state, action: PayloadAction<number>) => {
      if (action.payload) {
        state.posts = state.posts.filter((post) => post.id !== action.payload);
      }
      state.isLoading = false;
    });
    addCase(updatePostThunk.fulfilled, (state, action: PayloadAction<PostData>) => {
      const index = state.posts.findIndex((post) => post.id === action.payload.id);

      if (index !== -1) {
        state.posts[index] = action.payload;
      }
      state.isLoading = false;
    });

    addMatcher(
      isAnyOf(
        fetchPostsByUserIdThunk.pending,
        deletePostByIdThunk.pending,
        updatePostThunk.pending,
      ),
      (state) => {
        state.isLoading = true;
      },
    );
    addMatcher(
      isAnyOf(
        fetchPostsByUserIdThunk.rejected,
        deletePostByIdThunk.rejected,
        updatePostThunk.rejected,
      ),
      (state) => {
        state.isLoading = false;
      },
    );
  },
});

export default postsSlice.reducer;
