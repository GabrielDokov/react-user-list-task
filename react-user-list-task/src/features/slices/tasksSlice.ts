import { createSlice, isAnyOf, PayloadAction } from "@reduxjs/toolkit";
import { TaskData } from "../../types/TaskData";
import { fetchTasksThunk } from "../thunks/fetchTasksThunk";

export type TasksState = {
  tasks: TaskData[];
  isLoading: boolean;
};

const initialState: TasksState = {
  tasks: [],
  isLoading: false,
};

const tasksSlice = createSlice({
  name: "tasks",
  initialState,
  reducers: {
    toggleTaskStatus: (state, action: PayloadAction<number>) => {
      const task = state.tasks.find((t) => t.id === action.payload);
      if (task) {
        task.completed = !task.completed;
      }
    },
  },
  extraReducers: ({ addCase, addMatcher }) => {
    addCase(fetchTasksThunk.fulfilled, (state, action: PayloadAction<TaskData[]>) => {
      state.tasks = action.payload;
      state.isLoading = false;
    });

    addMatcher(isAnyOf(fetchTasksThunk.pending), (state) => {
      state.isLoading = true;
    });
    addMatcher(isAnyOf(fetchTasksThunk.rejected), (state) => {
      state.isLoading = false;
    });
  },
});

export const { toggleTaskStatus } = tasksSlice.actions;

export default tasksSlice.reducer;
