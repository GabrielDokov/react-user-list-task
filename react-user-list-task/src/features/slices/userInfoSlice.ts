import { createSlice, isAnyOf, PayloadAction } from "@reduxjs/toolkit";
import { fetchUsersThunk } from "../thunks/fetchUsersThunk";
import { UserData } from "../../types/UserData";

export type UsersState = {
  users: UserData[];
  isLoading: boolean;
};

const initialState: UsersState = {
  users: [],
  isLoading: false,
};

const userInfoSlice = createSlice({
  name: "userInfo",
  initialState,
  reducers: {
    updateUser: (state, action: PayloadAction<UserData>) => {
      const index = state.users.findIndex((user) => user.id === action.payload.id);
      if (index !== -1) {
        state.users[index] = action.payload;
      }
    },
  },
  extraReducers: ({ addCase, addMatcher }) => {
    addCase(fetchUsersThunk.fulfilled, (state, action: PayloadAction<UserData[]>) => {
      state.users = action.payload;
      state.isLoading = false;
    });

    addMatcher(isAnyOf(fetchUsersThunk.pending), (state) => {
      state.isLoading = true;
    });
    addMatcher(isAnyOf(fetchUsersThunk.rejected), (state) => {
      state.isLoading = false;
    });
  },
});

export const { updateUser } = userInfoSlice.actions;

export default userInfoSlice.reducer;
