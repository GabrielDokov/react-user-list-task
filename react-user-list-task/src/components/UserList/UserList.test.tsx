import UserList from "./UserList";
import renderWithProviders from "../../renderWithProviders";
import { usersMock } from "../../mocks/userMock";
import { UsersState } from "../../features/slices/userInfoSlice";

jest.mock("react-router", () => ({
  useNavigate: () => jest.fn(),
}));

const userInfo: UsersState = {
  users: usersMock,
  isLoading: false,
};

describe("components > UserList", () => {
  it("should match the snapshot with no data", async () => {
    const { asFragment } = renderWithProviders(<UserList />, {
      preloadedState: { userInfo },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with data from redux", async () => {
    const { asFragment } = renderWithProviders(<UserList />, {
      preloadedState: { userInfo },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
