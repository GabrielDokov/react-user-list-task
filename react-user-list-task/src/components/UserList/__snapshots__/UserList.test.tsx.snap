// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > UserList should match the snapshot with data from redux 1`] = `
<DocumentFragment>
  <h1
    class="ant-typography title css-dev-only-do-not-override-1vjf2v5"
  >
    Users
  </h1>
  <div
    class="ant-collapse ant-collapse-icon-position-start css-dev-only-do-not-override-1vjf2v5"
    role="tablist"
    style="width: 800px;"
  >
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="tab"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="collapsed"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          Leanne Graham
        </span>
      </div>
    </div>
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="tab"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="collapsed"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          TEst User
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > UserList should match the snapshot with no data 1`] = `
<DocumentFragment>
  <h1
    class="ant-typography title css-dev-only-do-not-override-1vjf2v5"
  >
    Users
  </h1>
  <div
    class="ant-collapse ant-collapse-icon-position-start css-dev-only-do-not-override-1vjf2v5"
    role="tablist"
    style="width: 800px;"
  >
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="tab"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="collapsed"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          Leanne Graham
        </span>
      </div>
    </div>
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="tab"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="collapsed"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          TEst User
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
